src/StringTools.cpp

```diff
@@ -0,0 +1,617 @@
+
+    Boole IsLowerAlphaLetter(const Char8 ToCheck)
+        { return ( ToCheck >= 'a' && ToCheck <= 'z' ); }
+
+    Boole IsUpperAlphaLetter(const Char8 ToCheck)
+        { return ( ToCheck >= 'A' && ToCheck <= 'Z' ); }
+
+    Boole IsAlphaLetter(const Char8 ToCheck)
+        { return ( IsLowerAlphaLetter(ToCheck) || IsUpperAlphaLetter(ToCheck) ); }
+
+    Boole IsLowerHexLetter(const Char8 ToCheck)
+        { return ( ToCheck >= 'a' && ToCheck <= 'f' ); }
+
+    Boole IsUpperHexLetter(const Char8 ToCheck)
+        { return ( ToCheck >= 'A' && ToCheck <= 'F' ); }
+
+    Boole IsHexLetter(const Char8 ToCheck)
+        { return ( IsLowerHexLetter(ToCheck) || IsUpperHexLetter(ToCheck) ); }
+
+    Boole IsHexDigit(const Char8 ToCheck)
+        { return ( IsDigit(ToCheck) || IsHexLetter(ToCheck) ); }
+
+    Boole IsAlphanumeric(const Char8 ToCheck)
+        { return ( IsDigit(ToCheck) || IsAlphaLetter(ToCheck) ); }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // String Manipulation and Checks
+
+    void ToUpperCase(StrIter Begin, StrIter End)
+    {
+        std::transform(Begin,End,Begin,[](int ToConvert) -> char {
+            return static_cast<char>( ::toupper(ToConvert) );
+        });
+    }
+
+    String UpperCaseCopy(String Source)
+    {
+        StringTools::ToUpperCase(Source.begin(),Source.end());
+        return Source;
+    }
+
+    void ToLowerCase(StrIter Begin, StrIter End)
+    {
+        std::transform(Begin,End,Begin,[](int ToConvert) -> char {
+            return static_cast<char>( ::tolower(ToConvert) );
+        });
+    }
+
+    String LowerCaseCopy(String Source)
+    {
+        StringTools::ToLowerCase(Source.begin(),Source.end());
+        return Source;
+    }
+
+    void ToCamelCase(StrIter Begin, StrIter End)
+    {
+        Boole PrevCharIsWhitespace = true;
+        for( StrIter CurrIt = Begin ; CurrIt != End ; ++CurrIt )
+        {
+            Boole CurrCharIsLowerLetter = StringTools::IsLowerAlphaLetter( *CurrIt );
+            Boole CurrCharIsUpperLetter = StringTools::IsUpperAlphaLetter( *CurrIt );
+            if( PrevCharIsWhitespace && CurrCharIsLowerLetter ) {
+                (*CurrIt) = static_cast<Char8>( ::toupper(*CurrIt) );
+            }else if( !PrevCharIsWhitespace && CurrCharIsUpperLetter ) {
+                (*CurrIt) = static_cast<Char8>( ::tolower(*CurrIt) );
+            }
+            PrevCharIsWhitespace = StringTools::IsWhitespace( *CurrIt );
+        }
+    }
+
+    String CamelCaseCopy(String Source)
+    {
+        StringTools::ToCamelCase(Source.begin(),Source.end());
+        return Source;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Size Modifying String Manipulations
+
+    void Trim(String& Source, const Boole Left, const Boole Right)
+    {
+        const String Delims = " \t\r";
+        if( Left ) {
+            size_t Forward = 0;
+            while( Forward < Source.size() )
```

Comment:
This doesn't seem like a reasonable implementation. Why not use std::find_first_of or std::find_if_not to find the range for std::erase.

Also, this is just conjecture, but I cannot imagine that the branching helps with optimization, if this were just a call RightTrim and LeftTrim then the game dev could decide which of the 3 to call and just never branch.

Reply:
The implementation has been updated as per a previous review comment.